tools/javadoc/javadoc.bzl

```
@@ -61,76 +58,53 @@ def _javadoc_library(ctx):
         # 1. Find the first directory under the working directory named '*java'.
         # 2. Assume all files to document can be found by appending a root_package name
         #    to that directory, or a subdirectory, replacting dots with slashes.
-        jargs.add_all(ctx.attr.root_packages)
-        jargs.add_joined("-subpackages", ctx.attr.root_packages, join_with = ":")
+        args.add_all(ctx.attr.root_packages)
+        args.add_joined("-subpackages", ctx.attr.root_packages, join_with = java_pathsep)
     else:
         # Document exactly the code in the specified source files.
-        jargs.add_all(ctx.files.srcs, map_each = _endswith_java)
+        args.add_all(ctx.files.srcs, map_each = _endswith_java)
 
     if ctx.attr.doctitle:
-        jargs.add("-doctitle", ctx.attr.doctitle)
+        args.add("-doctitle", ctx.attr.doctitle)
 
     # Translate `groups` mapping to `-group k ":".join(v)`.
-    jargs.add_all(
-        ctx.attr.groups.items(),
-        map_each = _to_group,
-    )
+    for group, packages in sorted(ctx.attr.groups.items()):
+        args.add("-group", group)
+        args.add_joined(packages, join_with = java_pathsep)
 
-    jargs.add_joined("-exclude", ctx.attr.exclude_packages, join_with = ":")
+    args.add_joined("-exclude", ctx.attr.exclude_packages, join_with = java_pathsep)
 
     for link in ctx.attr.external_javadoc_links:
-        jargs.add("-linkoffline {0} {0}".format(link))
+        args.add_all("-linkoffline", [link, link])
 
     if ctx.attr.bottom_text:
-        jargs.add("-bottom", ctx.attr.bottom_text)
+        args.add("-bottom", ctx.attr.bottom_text)
 
     srcs = depset(transitive = [src.files for src in ctx.attr.srcs]).to_list()
 
     # Invoke `javadoc` to generate Javadoc.
     ctx.actions.run_shell(
         inputs = srcs + classpath + ctx.files._jdk,
         outputs = [output_dir],
-        arguments = [
-            output_dir.path,
-            "true" if ctx.attr.root_packages else "false",
-            jargs,
-        ],
-        command = """
-outdir=$1; shift
-root_packages=$1; shift
-jargsfile=$1; shift
-
-# See also `TODO(b/167433657)` in javadoc.bzl.
-if $root_packages; then
-    sourcepath_args=(
-        -sourcepath
-        $(find * -type d -name '*.java' -print0 | tr '\\0' :)
-    )
-fi
-
-${JAVA_HOME}/bin/javadoc $jargsfile ${sourcepath_args[@]}
-""",
-        env = {
-            "JAVA_HOME": java_home,
-        },
+        arguments = [args],
+        command = "${JAVA_HOME}/bin/javadoc $1" + (
+            # See TODO(b/167433657) above.
+            ' -sourcepath "$(find * -type d -name \'*java\' -print0 | tr \'\\0\' :)"' if ctx.attr.root_packages else ""
```

```suggestion
            """ -sourcepath "$(find * -type d -name '*java' -print0 | tr '\\0' :)"""" if ctx.attr.root_packages else ""
```

Does this work? A little less backslash-escaping.